### DISCUSSION TITLE
Creating baseline from cli and adding the generated CurrentIssues as ManuallySupressedIssues

### DISCUSSION THREAD

**Discussion by laaptu** [detekt/detekt#4069] [Q&A]
Hi, I am trying to do the following
   * Create baseline from cli with 
```
configurations {
    detekt
}

task detektBaseline(type: JavaExec) {
    main = "io.gitlab.arturbosch.detekt.cli.Main"
    classpath = configurations.detekt
    def input = "$projectDir"
    def config = "detekt/detekt-config.yml"
    def baseline = "detekt/baseline.xml"
    def exclude = "**/build/**"
    def params = ['-i', input, '-c', config, '-ex', exclude, '-b', baseline,'-cb', '--build-upon-default-config']
    args(params)
}

```
* Then the generated `IDS` under `CurrentIssues`, I put inside `ManuallySuppressedIssues`. Meaning I simply copy  all the nodes of `CurrentIssues` and paste inside `ManuallySuppressedIssues` in the `baseline.xml`. As I want to ignore all the existing detekt issues.
* And when I run the detekt using 
```
task detekt(type: JavaExec) {
    main = "io.gitlab.arturbosch.detekt.cli.Main"
    classpath = configurations.detekt
    def input = "$projectDir"
    def config = "detekt/detekt-config.yml"
    def baseline = "detekt/baseline.xml"
    def exclude = "**/build/**"
    def params = ['-i', input, '-c', config, '-ex', exclude, '-b', baseline, '-ac']
    args(params)
}
```
I still get errors and the auto correction kicks in. My assumption was that once I add those `ID` into `ManuallySuppressedIssues`,  detekt won't complain. Is it the case that my `detektBaseline` is not able to find all the issues or is there anything I am doing wrong out here? Thank you.

**Comment by cortinico** [comment]
> * Meaning I simply copy  all the nodes of `CurrentIssues` and paste inside `ManuallySuppressedIssues` in the `baseline.xml`

That's totally fine, as `CurrentIssues` and `ManuallySuppressedIssues` are substancially equivalent. I actually have a couple of questions for you:

- Why you're not usign the Detekt Gradle Plugin (just out of curiosity if something is not working for you or you were not aware of it)?
- Seems like you added `'--build-upon-default-config'` only to the first task. Have you tried adding it also to the second (they should both either have it or not).
- What happens if you run `detektBaseline` once you moved the findings from `CurrentIssues` to `ManuallySuppressedIssues`

**Reply by laaptu** [reply]
@cortinico : Thank you for the reply and big thanks for making such an awesome tool. For the questions you have asked, please look upon my answers inline

* Why you're not usign the Detekt Gradle Plugin (just out of curiosity if something is not working for you or you were not aware of it)?
  * The current project was using `Detekt CLI` and I felt for adding baseline, I will simply go with the same CLI mechanism. The other reason is that I assumed for my given task `Detekt CLI` will be sufficient enough. But, I will definitely test with Detekt Gradle Plugin today to see whether it will be able to solve my needs.
  
* Seems like you added '--build-upon-default-config' only to the first task. Have you tried adding it also to the second (they should both either have it or not).
  * I tried all the combination with it (including the suggestions you have given here ) but couldn't see any difference. But I will again definitely try this today once again.
  
* What happens if you run detektBaseline once you moved the findings from CurrentIssues to ManuallySuppressedIssues
  * It will again create the same issues on the `baseline.xml`. Meaning let's say  after running `detektBaseline` I have 10 nodes under `CurrentIssues`. Then, I cut and paste those nodes to `ManuallySuppressedIssues`. And rerun `detektBaseline` again, then same 10 nodes appear on `CurrentIssues`. This has really confused me and I am bit doubtful about my implementation here.
  
  Thank you.

**Reply by cortinico** [reply]
> Then, I cut and paste those nodes to ManuallySuppressedIssues. And rerun detektBaseline again, then same 10 nodes appear on CurrentIssues

This to me sounds like a bug.
Can you please try point 2. and let us know if that was the issue first.
Then for this other problem, if you could create a reproducer, we should open an issue to treat it as a bug.

**Reply by laaptu** [reply]
@cortinico : Thank you. I tried the point2 without  `--build-upon-default-config` on both of the task, but still I am getting the same issue.
For the other problem, I tried with entirely new project ( https://github.com/laaptu/detekttest) and able to reproduce the issue. Have attached a video about it as well. 
Does this mean `Then for this other problem, if you could create a reproducer`, am I able to reproduce it again or not? Further, shall I try the same with older version of detekt and see whether I get the same issue or not? Thank you.

https://user-images.githubusercontent.com/97506/130543711-21e314d3-2930-4392-9196-0afb438221b1.mp4

**Reply by laaptu** [reply]
Further from my investigation, I found that if I don't use `-ac`, autocorrect option in my `detekt` command, then it will adhere to the `baseline.xml`  suppressed values. But if I use autocorrect option, then it will begin changes to the manually suppressed files. So, I guess, it is something to do with autocorrect option in the command line.

**Reply by BraisGabin** [reply]
Confirmed, I can reproduce this issue. Thanks for the video! It helped a lot!!

**Reply by laaptu** [reply]
@BraisGabin @cortinico  : Thank you for creating[ issue ticket ](https://github.com/detekt/detekt/issues/4073) for this. To get my understanding right, once this issue is resolved following things can be expected

- Once we add to `ManuallySuppressedIssues` in our baseline file , running `detektBaseline` will adhere to the baseline info. And  it will not  create a list of `CurrentIssues` identical to the `ManuallySuppressedIssues`  on the baseline file.
- When we run `detekt` from cli with `auto correction` on, it will adhere to the baseline file information and won't auto correct the files that are added to `ManuallySuppressedIssues` tags.

Thank you.

**Reply by cortinico** [reply]
> And it will not create a list of CurrentIssues identical to the ManuallySuppressedIssues on the baseline file.

That's correct

>  it will adhere to the baseline file information and won't auto correct the files that are added to ManuallySuppressedIssues tags.

Also correct 👍

**Reply by BraisGabin** [reply]
> > it will adhere to the baseline file information and won't auto correct the files that are added to ManuallySuppressedIssues tags.
> 
> Also correct 👍

Are we sure about this? As far as I know the baseline isn't related with the autoformat. I think that the autoformat happens before the baseline is parsed. So, even if you add an issue in `ManuallySuppressedIssues` the auto correct will fix it.

**Reply by laaptu** [reply]
> Are we sure about this? As far as I know the baseline isn't related with the autoformat. I think that the autoformat happens before the baseline is parsed. So, even if you add an issue in `ManuallySuppressedIssues` the auto correct will fix it.

@BraisGabin : Maybe I had wrong mental model about this. What I had assumed is that whatever we write on the `baseline.xml` will be ignored by detekt and will work on reporting issues that are not present on the baseline file. So based upon this, I assumed that detekt running with `auto-correct` will not begin to modify the files that has been added to baseline file. The reason I am bringing this up is that in our current project we are trying to migrate to  the new version of detekt and we would like to baseline some of the legacy files that do not adhere to our ` detekt config` rules. As we don't want to manually make correction to those files. So, for this case, what should be the right approach in detekt? Thank you.

**Reply by BraisGabin** [reply]
I'm not 100% sure of what I said. But as far as I know the baseline disables the issues in the report. But it doesn't disable the auto correct. But that should be easy to test. Other option is to use ktlint directly. They do have baseline too and I assume that they do support this feature that you are talking.

**Reply by laaptu** [reply]
Thank you @BraisGabin. I will look into the options you have suggested here. Thank you.
